src/ds/contiguous_set.h

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include <numeric>
+#include <vector>
+
+namespace ds
+{
+  // Dense representation of an ordered set of values, assuming it contains
+  // some contiguous ranges of adjacent values. Stores a sequence of ranges,
+  // rather than individual values.
+  template <typename T>
+  class ContiguousSet
+  {
+  public:
+    // Ranges are represented by their first value, and a count of additional
+    // values. This disallows negative ranges
+    using Range = std::pair<T, size_t>;
+    using Ranges = std::vector<Range>;
+
+    // Define an iterator for accessing each contained element, rather than the
+    // ranges
+    template <typename RangeIt>
+    struct TIterator
+    {
+      RangeIt it;
+      size_t offset = 0;
+
+      // clang-format off
```

Comment:
why?

Reply:
Was trying to keep this dense when there were a lot of methods. It's small enough now, so I've restored the standard formatting.